plugins/module_utils/firewall_rules_utils.py

```diff
@@ -437,40 +551,97 @@ def _load_rules(self) -> List[FirewallRule]:
 
     @property
     def changed(self) -> bool:
+        """
+        Checks if the current set of firewall rules has changed compared to the loaded configuration.
+
+        This property compares the current set of `FirewallRule` objects in `_rules` with the set
+        loaded from the configuration file. It returns True if there are differences, indicating
+        that changes have been made to the ruleset which are not yet saved to the configuration file.
+
+        Returns:
+            bool: True if the ruleset has changed, False otherwise.
+        """
         return self._load_rules() != self._rules
 
     def add_or_update(self, rule: FirewallRule) -> None:
-        existing_rule: Optional[FirewallRule] = next(
-            (r for r in self._rules if r == rule), None
-        )
+        """
+        Adds a new firewall rule to the ruleset or updates an existing one.
+
+        This method checks if the provided `rule` already exists in the ruleset. If it does,
+        the existing rule is updated with the properties of the provided `rule`. If it does not exist,
+        the new rule is appended to the ruleset. The comparison to check if a rule exists is based on
+        the equality condition defined in the `FirewallRule` class.
+
+        Parameters:
+            rule (FirewallRule): The firewall rule to be added or updated in the ruleset.
+
+        Returns:
+            None: This method does not return anything.
+        """
+
+        existing_rule: Optional[FirewallRule] = next((r for r in self._rules if r == rule), None)
         if existing_rule:
             existing_rule.__dict__.update(rule.__dict__)
         else:
             self._rules.append(rule)
 
     def delete(self, rule: FirewallRule) -> None:
+        """
+        Removes a specified firewall rule from the ruleset.
+
+        This method iterates through the current set of firewall rules and removes the rule
+        that matches the provided `rule` parameter. The comparison for removal is based on
+        the inequality of the `FirewallRule` objects. If the rule is not found, no action is taken.
+
+        Parameters:
+            rule (FirewallRule): The firewall rule to be removed from the ruleset.
+
+        Returns:
+            None: This method does not return anything.
+        """
+
         self._rules = [r for r in self._rules if r != rule]
```

Comment:
unrelated, but makes more sense to use built in functions
```suggestion
        self._rules.remove(rule)